### PR TITLE
Add `tokenize_batch` method

### PR DESCRIPTION
**Pull request by michael-p**
The Python bindings provide this method as a convenience function. I think this is useful for users of the Rust library as well, so I implemented it there. This means we have to take on a dependency on `ndarray`, but I think this is justified given how widely used this crate is in the ML ecosystem.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -189,6 +191,58 @@ impl Tokenizer {
...
+    pub fn tokenize_batch<'a, I>(&self, texts: I, context_length: usize) -> Array2<u16>
+    where
+        I: IntoIterator<Item = &'a str>,
+        I::IntoIter: ExactSizeIterator,
+    {
+        if context_length < 3 {
+            panic!("context length must be at least 3");
+        }
+        let texts = texts.into_iter();
+        let mut result = Array2::zeros((texts.len(), context_length));
```
Comment (djc) [inline src/lib.rs:231]: If we're only using ndarray for `Array2` and the ability to iterate over it, it feels like we can make a minimal take on that without pulling in the large dependency?

Alternatively we should at least gate this with a `feature` gate and make the ndarray dependency optional.

File: src/lib.rs [hunk 2]
```diff
@@ -189,6 +191,58 @@ impl Tokenizer {
...
+    /// # use ndarray::array;
+    /// # use instant_clip_tokenizer::{Token, Tokenizer};
+    /// let tokenizer = Tokenizer::new();
+    /// let encoded = tokenizer.tokenize_batch(["Hi", "How are you?"], 5);
+    /// assert_eq!(encoded, array![
+    ///     [49406, 1883, 49407, 0, 0],
+    ///     [49406, 829, 631, 592, 49407],
+    /// ]);
+    /// ```
+    pub fn tokenize_batch<'a, I>(&self, texts: I, context_length: usize) -> Array2<u16>
```
Comment (djc) [inline src/lib.rs:222]: I prefer to use `impl IntoIterator<Item = &'a str>` over separate bounds where possible, since this saves the indirection of a type argument name.

File: src/lib.rs [hunk 3]
```diff
@@ -189,6 +191,58 @@ impl Tokenizer {
...
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use ndarray::array;
+    /// # use instant_clip_tokenizer::{Token, Tokenizer};
+    /// let tokenizer = Tokenizer::new();
+    /// let encoded = tokenizer.tokenize_batch(["Hi", "How are you?"], 5);
+    /// assert_eq!(encoded, array![
+    ///     [49406, 1883, 49407, 0, 0],
```
Comment (djc) [inline src/lib.rs:218]: This makes it seem like the notion of trailing zeros is somehow specialized. Is that the case? How/where is that case handled?

**Review by djc — COMMENTED** [review]
I have some suggestions!

File: src/lib.rs (hunk 1 above)
Comment (michael-p) [reply src/lib.rs:231]: We need `Array2` as the return type. We could of course return a `Vec<u16>` instead with just the data, but that would put the burden on the caller to construct some matrix type from that (be it `ndarray::Array2` or `nalgebra::Matrix`).
For instance, the caller would need to think about the shape of the matrix, and whether the data is row-major or column-major order. Hence, returning a specific type representing a matrix is a lot more convenient and less error-prone for the caller.
We could use a different matrix type (e.g. from the `nalgebra` crate, or even roll our own with conversions to/from `ndarray`, ...) but `ndarray` seems what basically everyone is using in the context of ML so I chose that.

In summary, I would prefer to keep it that way, gated by a crate feature.

File: src/lib.rs (hunk 2 above)
Comment (michael-p) [reply src/lib.rs:222]: > I prefer to use impl IntoIterator<Item = &'a str> over separate bounds where possible, since this saves the indirection of a type argument name.

I'm not sure how this can be done, given that we also need the `I::IntoIter: ExactSizeIterator` bound?

> Also, I think we should name this to match the other methods, so probably encode_batch()?

I intentionally chose a different name because it does a bit more than `encode`, namely add `<start_of_text>` and  `<end_of_text>` markers, and truncate to `context_length`.
The names `tokenize` and `encode` are also what the original Python library uses for the corresponding functions.

Comment (djc) [reply src/lib.rs:222]: > > I prefer to use impl IntoIterator<Item = &'a str> over separate bounds where possible, since this saves the indirection of a type argument name.
> 
> I'm not sure how this can be done, given that we also need the `I::IntoIter: ExactSizeIterator` bound?

Ah, that's fair -- sorry for the noise.

> > Also, I think we should name this to match the other methods, so probably encode_batch()?
> 
> I intentionally chose a different name because it does a bit more than `encode`, namely add `<start_of_text>` and `<end_of_text>` markers, and truncate to `context_length`. The names `tokenize` and `encode` are also what the original Python library uses for the corresponding functions.

Alright. Just make sure the difference is very clear in the documentation.

File: src/lib.rs (hunk 3 above)
Comment (djc) [reply src/lib.rs:218]: > Now, in case the tokenization of an input string results in less than `context_length` tokens the remaining matrix elements need to have some value, and `0` is what the Python code uses as a "padding".

I was talking about this input value stuff. IIRC there is no other case in the Rust library where this padding concept is used, so we should make sure it is adequately documented here. We're building a Rust library here so it's not sufficient to implicitly assume choices from another Python library in the public interface of this library.

File: src/lib.rs (hunk 1 above)
Comment (djc) [reply src/lib.rs:231]: Fair enough.

File: Cargo.toml [hunk 4]
```diff
@@ -15,12 +15,13 @@ required-features = ["openai-vocabulary-file"]
...
+ndarray = { version = "0.15.6", optional = true }
 regex = "1.10.2"
 
 [dev-dependencies]
 criterion = "0.5.1"
 
 [[bench]]
 name = "bench"
-required-features = ["openai-vocabulary-file"]
+required-features = ["ndarray", "openai-vocabulary-file"]
```
Comment (djc) [inline Cargo.toml:26]: We should avoid this and only guard the `tokenize_batch_small` benchmark on this.

File: src/lib.rs [hunk 5]
```diff
@@ -41,15 +41,21 @@
...
-//! bundles the default vocabulary file used for OpenAI's CLIP model together
-//! with this crate and allows users to construct a new tokenizer simply by
-//! calling [`Tokenizer::new`].
+//! This crate provides two features, both of which are enabled by default:
 //!
-//! This feature is enabled by default. To disable it use `default-features =
-//! false` when specifying the dependency on this crate in your `Cargo.toml`.
-//! You will need to supply your own vocabulary file then and construct the
-//! tokenizer using [`Tokenizer::with_vocabulary`].
+//! * **ndarray** - Adds a dependency on the
```
Comment (djc) [inline src/lib.rs:46]: You write "Adds [..] and enabled [..]". Should probably be consistent and use "enables".

File: src/lib.rs [hunk 6]
```diff
@@ -41,15 +41,21 @@
...
+//!   [`ndarray`](https://docs.rs/ndarray) crate and enabled the
+//!  [`Tokenizer::tokenize_batch`] method that can be used to tokenize several
+//!  input strings at once and returns a matrix suitable for directly passing to
+//!  the CLIP neural network.
```
Comment (djc) [inline src/lib.rs:50]: The "can be used to [..] at once and returns" feels a little awkward. How about, "at once, returning"?

File: src/lib.rs [hunk 7]
```diff
@@ -41,15 +41,21 @@
...
+//! * **openai-vocabulary-file** - This feature bundles the default vocabulary
+//!   file used for OpenAI's CLIP model together with this crate and allows
+//!   users to construct a new tokenizer simply by calling [`Tokenizer::new`].
+//!   When disabled, you will need to supply your own vocabulary file then and
```
Comment (djc) [inline src/lib.rs:54]: Remove "then"?

File: Cargo.toml [hunk 8]
```diff
@@ -6,7 +6,7 @@ rust-version = "1.65"
 license = "MIT"
 
 [features]
-default = ["openai-vocabulary-file"]
+default = ["ndarray", "openai-vocabulary-file"]
```
Comment (djc) [inline Cargo.toml:9]: Hmm, it feels like `ndarray` should not be a part of `default`.

File: src/lib.rs [hunk 9]
```diff
@@ -189,16 +195,78 @@ impl Tokenizer {
...
+    /// numeric representation written to a row in the resulting two-dimensional
+    /// matrix of shape `(texts.len(), context_length)`, with the special
+    /// `<start_of_text>` token prepended, and `<end_of_text>` appended to each
+    /// text.
+    ///
+    /// The individual input strings are lowercased before being tokenized, but
+    /// otherwise no pre-processing is performed.
+    ///
+    /// `context_length` is the maximum number of tokens per each text and
+    /// should be `77` for all current CLIP models. If tokenization results less
```
Comment (djc) [inline src/lib.rs:210]: "results" -> add "in"?

File: src/lib.rs [hunk 10]
```diff
@@ -189,16 +195,78 @@ impl Tokenizer {
...
+    /// than `context_length` tokens the resulting row will be padded with
+    /// trailing zeros. If tokenizing an input text results in too many tokens,
+    /// the token sequence will be truncated to fit within the resulting row of
+    /// length `context_length`, always including the `<start_of_text>` and
+    /// `<end_of_text>` marker tokens.
+    ///
+    /// The resulting matrix can be passed directly to the CLIP neural network.
```
Comment (djc) [inline src/lib.rs:217]: Should add a `# Panics` section.

File: Cargo.toml (hunk 4 above)
Comment (djc) [reply Cargo.toml:31]: Why is this better than just guarding the single benchmark?

**Review by djc — APPROVED** [review]

File: Cargo.toml (hunk 4 above)
Comment (michael-p) [reply Cargo.toml:31]: I'm not sure guarding a single benchmark is even possible in Criterion. I didn't find anything in the documentation, and tried several combinations while defining the `criterion_group!(...)` and `criterion_main!(...)` but it looks like there is no way to feature-gate individual benchmarks.

In addition, it's maybe a bit more discoverable. I tend to just run `cargo bench` on projects I'm not familiar with (instead of `cargo bench --all-features`) and would hence miss those benchmarks. By having two benchmark `.rs` files I might realize that there are actually more benchmarks I can run.
